server/src/main/java/io/crate/expression/symbol/GroupAndAggregateSemantics.java

```diff
@@ -76,6 +79,40 @@ public static void validate(List<Symbol> outputSymbols,
         }
     }
 
+    private static void validateGroupBySymbol(Symbol groupBy) {
+        groupBy.accept(SymbolTypeValidator.INSTANCE, null);
+    }
+
+    private static class SymbolTypeValidator extends SymbolVisitor<Void, Void> {
+
+        static final SymbolTypeValidator INSTANCE = new SymbolTypeValidator();
+
+        @Override
+        public Void visitSymbol(Symbol symbol, Void context) {
+            if (symbol.valueType() == DataTypes.UNDEFINED) {
+                throw new IllegalArgumentException(
+                    "Cannot group or aggregate on '" + symbol.toString() + "' efficiently with an undefined type." +
+                    " Use an explicit type cast."
```

Comment:
I think the error is a bit misleading. It's not that it cannot do it efficiently without a defined type, it can't do it at all.

It works with casts, but the cast has a bit of overhead and given that it's most likely to be undefined because of being an ignored column it won't have a column-store (docvalues) available and it will use a source lookup and other more expensive code paths.